core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -284,8 +284,24 @@ RexNode simplify(RexNode e, RexUnknownAs unknownAs) {
     case NOT_EQUALS:
       return simplifyComparison((RexCall) e, unknownAs);
     default:
+      if (e.getClass() == RexCall.class) {
+        return simplifyGenericNode((RexCall) e);
```

Comment:
Suppose there's a generic `f(x)`.
Then `f(f(f(f(f(x)))))` would seem to have `O(N^2)` complexity as each time it would recurse all the way down.

Any thoughts on that?

Reply:
`simplifyGenericNode` runs simplification only on the call operands.

on the following testcase the code does 2047 rexnode visits ; tree is composed from 2047 nodes
```
  @Test public void testSimplifyRecurseIntoArithmetics2() {
    RexNode n = literal(3);
    for (int i = 0; i < 10; i++) {
      n = plus(n, n);
    }

    checkSimplify(
        n,
        "+(1, 2)");
  }
```

...or is there something I miss here?
Do you have a testcase for this?

Reply:
no worries, thank you for taking a look!
I've opened a ticket to add something which could catch these revisitation issues CALCITE-2851

Reply:
CALCITE-2851 is a nice ticket.
I wonder if we could add an automatic check there: for instance, fail the test when number of simplify actions exceeds 3*N (or whatever).

Reply:
+1

Reply:
...and, if and when we have that check, we should welcome bug reports of conditions that breach that check. We might discover, for instance, that '- - - - - - - - - - x' requires 1024 (2 ^ 10) simplify operations. That would be a bug. But we might decide that it is a low priority bug, and not worth fixing.

Reply:
I really hope that we could keep an upper bound to visit at most every node twice even in the worst case - but I might be wrong